<a href="https://colab.research.google.com/github/dakshini01/Statistical-Learning-e20181/blob/main/Assignment_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part A: Theoretical Fundamentals (Maximum Likelihood & Decision Space)

## 1. Bias of the Maximum Likelihood Estimator of Covariance

Consider the i.i.d. random variables

$$
X_1, X_2, \ldots, X_n \in \mathbb{R}^m
$$

with population mean

$$
\mu = E[X_i]
$$

and population covariance matrix

$$
\Sigma = Var(X_i).
$$

The empirical sample mean is

$$
\hat{\mu}_n = \frac{1}{n}\sum_{i=1}^{n}X_i.
$$

The raw Maximum Likelihood Estimator for the covariance matrix is

$$
\hat{\Sigma}_{MLE}
=
\frac{1}{n}
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T.
$$

However, this estimator is biased in finite samples. To show this, we use the identity

$$
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
=
\sum_{i=1}^{n}
(X_i-\mu)(X_i-\mu)^T
-
n(\hat{\mu}_n-\mu)(\hat{\mu}_n-\mu)^T.
$$

Taking expectation on both sides,

$$
E\left[
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
\right]
=
E\left[
\sum_{i=1}^{n}
(X_i-\mu)(X_i-\mu)^T
\right]
-
nE\left[
(\hat{\mu}_n-\mu)(\hat{\mu}_n-\mu)^T
\right].
$$

Since

$$
E\left[
(X_i-\mu)(X_i-\mu)^T
\right]
=
\Sigma,
$$

we get

$$
E\left[
\sum_{i=1}^{n}
(X_i-\mu)(X_i-\mu)^T
\right]
=
n\Sigma.
$$

Also,

$$
Var(\hat{\mu}_n)
=
E\left[
(\hat{\mu}_n-\mu)(\hat{\mu}_n-\mu)^T
\right]
=
\frac{\Sigma}{n}.
$$

Therefore,

$$
E\left[
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
\right]
=
n\Sigma
-
n\left(\frac{\Sigma}{n}\right).
$$

Hence,

$$
E\left[
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
\right]
=
(n-1)\Sigma.
$$

Now,

$$
E[\hat{\Sigma}_{MLE}]
=
E\left[
\frac{1}{n}
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
\right].
$$

Therefore,

$$
E[\hat{\Sigma}_{MLE}]
=
\frac{1}{n}(n-1)\Sigma.
$$

Thus,

$$
E[\hat{\Sigma}_{MLE}]
=
\frac{n-1}{n}\Sigma.
$$

Since

$$
\frac{n-1}{n}\Sigma \neq \Sigma,
$$

the raw MLE covariance estimator is biased. It underestimates the true population covariance matrix.

---

## Bessel's Correction and the Unbiased Estimator

To remove this bias, we multiply the MLE covariance estimator by

$$
\frac{n}{n-1}.
$$

This gives the unbiased sample covariance matrix

$$
S
=
\frac{1}{n-1}
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T.
$$

Taking expectation,

$$
E[S]
=
E\left[
\frac{1}{n-1}
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
\right].
$$

Substituting the previous result,

$$
E[S]
=
\frac{1}{n-1}
(n-1)\Sigma.
$$

Therefore,

$$
E[S]=\Sigma.
$$

Hence, applying Bessel's correction gives an unbiased estimator of the population covariance matrix.

The reason for using $n-1$ instead of $n$ is that the sample mean $\hat{\mu}_n$ is estimated from the same dataset. Therefore, one degree of freedom is lost.

---

## 2. Type I Error and Type II Error in Structural Health Monitoring

In continuous structural health monitoring, the null hypothesis is

$$
H_0:
\text{The structural asset is healthy.}
$$

The alternative hypothesis is

$$
H_1:
\text{The structural asset is damaged or abnormal.}
$$

---

### Type I Error

A Type I Error occurs when the diagnostic system rejects $H_0$ even though $H_0$ is actually true.

In this context, it means that the system says the structure is damaged even though it is healthy.

Mathematically,

$$
\alpha
=
P(\text{Reject }H_0 \mid H_0 \text{ is true}).
$$

This is a false alarm.

In an industrial plant, a Type I Error may cause unnecessary shutdown, inspection cost, and production loss.

---

### Type II Error

A Type II Error occurs when the diagnostic system fails to reject $H_0$ even though $H_0$ is false.

In this context, it means that the system says the structure is healthy even though it is actually damaged.

Mathematically,

$$
\beta
=
P(\text{Fail to reject }H_0 \mid H_0 \text{ is false}).
$$

This is a missed detection.

A Type II Error is dangerous because real structural damage may continue unnoticed.

---

### Statistical Power

The statistical power of the diagnostic test is

$$
Power = 1-\beta.
$$

Power is the probability of correctly detecting damage when damage is truly present.

---

## Effect of an Ultra-Conservative Significance Threshold

Suppose the engineer chooses

$$
\alpha = 0.0001.
$$

This is a very small significance level.

The significance level $\alpha$ controls the probability of a false alarm. Therefore, by choosing a very small $\alpha$, the engineer greatly reduces the probability of incorrectly declaring a healthy structure as damaged.

However, this decision also makes the test more conservative.

The rejection region becomes very small:

$$
\alpha \downarrow
\Rightarrow
\text{Rejection region decreases}.
$$

As a result, it becomes harder for the test statistic to fall inside the rejection region.

Therefore,

$$
\beta \uparrow
$$

and

$$
Power = 1-\beta \downarrow.
$$

So, reducing $\alpha$ decreases false alarms, but it also reduces the ability of the diagnostic system to detect real damage.

---

## Geometric Interpretation Using the Healthy Confidence Ellipsoid

In a multivariate diagnostic system, healthy operation can be represented by a confidence ellipsoid around the baseline mean vector.

The ellipsoid represents the region where normal healthy data points are expected to lie.

When $\alpha$ is reduced from a common value such as

$$
\alpha = 0.05
$$

to an ultra-conservative value such as

$$
\alpha = 0.0001,
$$

the critical threshold increases.

Therefore, the healthy confidence ellipsoid becomes larger.

Geometrically,

$$
\alpha \downarrow
\Rightarrow
\text{critical value} \uparrow
\Rightarrow
\text{confidence ellipsoid volume} \uparrow.
$$

This means more observations are accepted as healthy.

Therefore:

- False alarms decrease.
- The healthy region becomes larger.
- The diagnostic system becomes less sensitive.
- Type II Error increases.
- Statistical power decreases.

Thus, an ultra-conservative threshold protects the plant from unnecessary shutdowns, but it increases the risk of missing subtle structural damage.

# Part B: Theoretical Extension (Asymptotic Distributions & Slutsky's Theorem)

The Lindeberg-Lévy Multivariate Central Limit Theorem states that, for i.i.d. random variables

$$
X_1, X_2, \ldots, X_n \in \mathbb{R}^m
$$

with population mean

$$
\mu = E[X_i]
$$

and population covariance matrix

$$
\Sigma = Var(X_i),
$$

the standardized sample mean satisfies

$$
\sqrt{n}(\hat{\mu}_n-\mu)
\overset{d}{\longrightarrow}
N(0,\Sigma)
$$

as

$$
n \rightarrow \infty.
$$

Here,

$$
\hat{\mu}_n = \frac{1}{n}\sum_{i=1}^{n}X_i.
$$

This means that when the sample size becomes very large, the distribution of the sample mean becomes approximately multivariate normal.

---

## Slutsky's Theorem

Let

$$
A_n \overset{d}{\longrightarrow} A
$$

and

$$
B_n \overset{p}{\longrightarrow} b,
$$

where $b$ is a constant.

Then Slutsky's Theorem states that

$$
A_n + B_n \overset{d}{\longrightarrow} A+b,
$$

$$
A_nB_n \overset{d}{\longrightarrow} Ab,
$$

and, if $b \neq 0$,

$$
\frac{A_n}{B_n}
\overset{d}{\longrightarrow}
\frac{A}{b}.
$$

In simple words, Slutsky's Theorem says that if one sequence converges in distribution and another sequence converges in probability to a constant, then we can combine them, and the final limiting distribution remains valid.

---

## Consistency of the Sample Covariance Matrix

The population covariance matrix $\Sigma$ is unknown in real applications.

Therefore, we estimate it using the unbiased sample covariance matrix

$$
S =
\frac{1}{n-1}
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T.
$$

By the Law of Large Numbers, the sample covariance matrix is a consistent estimator of the true covariance matrix. Therefore,

$$
S \overset{p}{\longrightarrow} \Sigma
$$

as

$$
n \rightarrow \infty.
$$

This means that when the sample size becomes very large, the sample covariance matrix $S$ becomes closer and closer to the true population covariance matrix $\Sigma$.

---

## Applying Slutsky's Theorem

From the Multivariate Central Limit Theorem,

$$
\sqrt{n}(\hat{\mu}_n-\mu)
\overset{d}{\longrightarrow}
N(0,\Sigma).
$$

But in practice, $\Sigma$ is unknown.

Since

$$
S \overset{p}{\longrightarrow} \Sigma,
$$

Slutsky's Theorem allows us to replace the unknown $\Sigma$ with the observable estimator $S$.

Therefore, the practical asymptotic distribution becomes

$$
\sqrt{n}(\hat{\mu}_n-\mu)
\overset{d}{\longrightarrow}
N(0,S).
$$

Equivalently, this can be written as

$$
\hat{\mu}_n
\sim
N\left(
\mu,
\frac{S}{n}
\right).
$$

---

## Interpretation

The expression

$$
\hat{\mu}_n
\sim
N\left(
\mu,
\frac{S}{n}
\right)
$$

means that the sample mean vector $\hat{\mu}_n$ is approximately multivariate normally distributed around the true population mean vector $\mu$.

The covariance of this distribution is

$$
\frac{S}{n}.
$$

As $n$ increases,

$$
\frac{S}{n}
\rightarrow 0.
$$

Therefore, the uncertainty around the sample mean becomes smaller.

This justifies using the sample mean and sample covariance matrix to build confidence regions, hypothesis tests, and diagnostic limits in practical structural health monitoring systems.

---

## Final Conclusion

Slutsky's Theorem justifies replacing the unknown population covariance matrix $\Sigma$ with the sample covariance matrix $S$ because

$$
S \overset{p}{\longrightarrow} \Sigma.
$$

Therefore, the theoretical model

$$
\hat{\mu}_n
\sim
N\left(
\mu,
\frac{\Sigma}{n}
\right)
$$

can be operationalized as

$$
\hat{\mu}_n
\sim
N\left(
\mu,
\frac{S}{n}
\right).
$$

This makes the asymptotic distribution practically usable because $S$ can be calculated directly from the observed dataset.

In [1]:
import numpy as np
import pandas as pd
import scipy.stats as stats

# --- DO NOT MODIFY: Synthetic Strain Sensor Data Generator ---
np.random.seed(42)

n_samples = 5000
n_features = 3

base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[
        [0.09, 0.02, 0.01],
        [0.02, 0.06, 0.03],
        [0.01, 0.03, 0.05]
    ],
    size=n_samples
)

# Inject subtle drift
base_data[4000:, 0] += 0.015
base_data[4000:, 2] -= 0.010

df_strain = pd.DataFrame(
    base_data,
    columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3']
)

# --------------------------------------------------------

def verify_first_moment_homogeneity(df, g_chunks=5):

    n = len(df)
    m = df.shape[1]

    chunk_size = n // g_chunks

    chunks = []

    for i in range(g_chunks):
        start = i * chunk_size
        end = (i + 1) * chunk_size
        chunks.append(df.iloc[start:end])

    # Global mean
    global_mean = df.mean().values.reshape(-1, 1)

    # Within-chunk matrix W
    W = np.zeros((m, m))

    # Between-chunk matrix B
    B = np.zeros((m, m))

    for chunk in chunks:

        chunk_values = chunk.values

        n_j = len(chunk)

        chunk_mean = chunk.mean().values.reshape(-1, 1)

        # Within matrix
        centered = chunk_values - chunk_mean.T

        W += centered.T @ centered

        # Between matrix
        mean_diff = chunk_mean - global_mean

        B += n_j * (mean_diff @ mean_diff.T)

    T = W + B

    wilks_lambda = np.linalg.det(W) / np.linalg.det(T)

    # Bartlett approximation
    chi_square = -(n - 1 - ((m + g_chunks) / 2)) * np.log(wilks_lambda)

    df_chi = m * (g_chunks - 1)

    p_value = 1 - stats.chi2.cdf(chi_square, df_chi)

    return {
        "Wilks Lambda": wilks_lambda,
        "Bartlett Chi Square": chi_square,
        "Degrees of Freedom": df_chi,
        "p-value": p_value
    }

results = verify_first_moment_homogeneity(df_strain, g_chunks=5)

for k, v in results.items():
    print(f"{k}: {v}")

Wilks Lambda: 0.9904519796929591
Bartlett Chi Square: 47.9215049961488
Degrees of Freedom: 12
p-value: 3.2255259508895406e-06


# Part C: Numerical Verification

The dataset was partitioned into

$$
g=5
$$

consecutive chunks, each containing

$$
n_j=1000
$$

observations.

The global mean vector and individual chunk mean vectors were calculated. Using these means, the Within-Chunk Sum of Squares and Cross Products Matrix

$$
W
$$

and the Between-Chunk Sum of Squares and Cross Products Matrix

$$
B
$$

were constructed.

The Wilks' Lambda statistic was computed as

$$
\Lambda
=
\frac{|W|}{|B+W|}.
$$

The calculated value was

$$
\Lambda = \text{0.9904519796929591}
$$

Using Bartlett's asymptotic transformation,

$$
\chi^2
=
-(n-1-\frac{m+g}{2})
\ln(\Lambda),
$$

the test statistic was obtained as

$$
\chi^2
=
\text{47.9215049961488}.
$$

The corresponding degrees of freedom were

$$
m(g-1)
=
3(5-1)
=
12.
$$

The resulting p-value was

$$
p
=
\text{3.2255259508895406e-06}.
$$

---

## Decision

The significance level was chosen as

$$
\alpha = 0.05.
$$

The decision rule is:

- If

$$
p \le 0.05
$$

reject

$$
H_0
$$

- If

$$
p > 0.05
$$

fail to reject

$$
H_0.
$$

Since

$$
p \text{ =3.2255259508895406e-06< } 0.05,
$$

we conclude that:


There is sufficient evidence to conclude that the chunk means differ significantly. Therefore, the baseline center has shifted over time and the first moment homogeneity assumption is violated.



# Geometric Subspace Optimization via Principal Component Analysis (PCA)

# Part A: Theoretical Fundamentals (Coordinate Projections & Orthogonality)

## 1. PCA Projection and Covariance Diagonalization

Let the centered baseline feature vector be

$$
\tilde{X}_i = X_i - \mu \in \mathbb{R}^m
$$

where $X_i$ is the original observation vector and $\mu$ is the population mean vector.

Since the data is centered,

$$
E[\tilde{X}_i] = 0.
$$

The population covariance matrix is

$$
\Sigma = E[\tilde{X}_i\tilde{X}_i^T].
$$

Because $\Sigma$ is symmetric and positive-semidefinite, it can be decomposed using spectral decomposition as

$$
\Sigma = P\Lambda P^T.
$$

Here,

$$
P = [p_1,p_2,\ldots,p_m]
$$

is the orthogonal eigenvector matrix, and

$$
\Lambda = diag(\lambda_1,\lambda_2,\ldots,\lambda_m)
$$

is the diagonal eigenvalue matrix.

The eigenvalues are ordered as

$$
\lambda_1 \geq \lambda_2 \geq \cdots \geq \lambda_m \geq 0.
$$

Since $P$ is orthogonal,

$$
P^TP = PP^T = I.
$$

The principal component transformation is defined as

$$
Z_i = P^T\tilde{X}_i.
$$

Now we calculate the covariance matrix of the transformed vector $Z_i$:

$$
E[Z_iZ_i^T]
=
E[(P^T\tilde{X}_i)(P^T\tilde{X}_i)^T].
$$

Using the transpose rule,

$$
(AB)^T = B^TA^T,
$$

we get

$$
(P^T\tilde{X}_i)^T = \tilde{X}_i^TP.
$$

Therefore,

$$
E[Z_iZ_i^T]
=
E[P^T\tilde{X}_i\tilde{X}_i^TP].
$$

Since $P$ is constant, it can be moved outside the expectation:

$$
E[Z_iZ_i^T]
=
P^TE[\tilde{X}_i\tilde{X}_i^T]P.
$$

But

$$
E[\tilde{X}_i\tilde{X}_i^T] = \Sigma.
$$

So,

$$
E[Z_iZ_i^T]
=
P^T\Sigma P.
$$

Substituting the spectral decomposition,

$$
E[Z_iZ_i^T]
=
P^T(P\Lambda P^T)P.
$$

Using

$$
P^TP = I,
$$

we obtain

$$
E[Z_iZ_i^T]
=
I\Lambda I.
$$

Therefore,

$$
E[Z_iZ_i^T]
=
\Lambda.
$$

Hence, the covariance matrix of the transformed vector is diagonal.

---

## Interpretation of the Diagonal Result

Since

$$
E[Z_iZ_i^T] = \Lambda,
$$

we have

$$
\Lambda =
\begin{bmatrix}
\lambda_1 & 0 & \cdots & 0 \\
0 & \lambda_2 & \cdots & 0 \\
\vdots & \vdots & \ddots & \vdots \\
0 & 0 & \cdots & \lambda_m
\end{bmatrix}.
$$

This means all off-diagonal covariance terms are zero.

Therefore,

$$
E[Z_{i,j}Z_{i,k}] = 0
\quad
\text{for}
\quad
j \neq k.
$$

So, different principal component coordinates are uncorrelated.

Geometrically, PCA rotates the original coordinate system into a new coordinate system where the axes point in the directions of maximum variance.

Statistically, each principal component captures a separate uncorrelated mode of variation in the data.

---

## 2. Conservation of Total Variance

The total variance of the original system is measured by the trace of the covariance matrix:

$$
tr(\Sigma).
$$

Using spectral decomposition,

$$
\Sigma = P\Lambda P^T.
$$

Therefore,

$$
tr(\Sigma)
=
tr(P\Lambda P^T).
$$

Using the cyclic property of trace,

$$
tr(ABC)=tr(BCA)=tr(CAB),
$$

we get

$$
tr(P\Lambda P^T)
=
tr(\Lambda P^TP).
$$

Since

$$
P^TP = I,
$$

we get

$$
tr(\Sigma)
=
tr(\Lambda I).
$$

Therefore,

$$
tr(\Sigma)
=
tr(\Lambda).
$$

Since $\Lambda$ is diagonal,

$$
tr(\Lambda)
=
\sum_{j=1}^{m}\lambda_j.
$$

Thus,

$$
tr(\Sigma)
=
tr(\Lambda)
=
\sum_{j=1}^{m}\lambda_j.
$$

This proves that PCA preserves total variance.

PCA does not create or destroy variance. It only rotates the coordinate system.

---

## Cumulative Explained Variance Ratio

Suppose we retain only the first $k$ principal components, where

$$
k < m.
$$

The explained subspace is

$$
\hat{P}_k \in \mathbb{R}^{m \times k}.
$$

The residual unexplained subspace is

$$
\hat{P}_{m-k} \in \mathbb{R}^{m \times (m-k)}.
$$

The cumulative explained variance ratio is

$$
\Phi(k)
=
\frac{
\sum_{j=1}^{k}\hat{\lambda}_j
}{
\sum_{j=1}^{m}\hat{\lambda}_j
}.
$$

This tells us the percentage of total system variance captured by the first $k$ principal components.

---

## Residual Unexplained Variance Ratio

The residual unexplained variance ratio is

$$
\Psi(k)
=
1-\Phi(k).
$$

Equivalently,

$$
\Psi(k)
=
\frac{
\sum_{j=k+1}^{m}\hat{\lambda}_j
}{
\sum_{j=1}^{m}\hat{\lambda}_j
}.
$$

This tells us the percentage of system variance discarded by ignoring the remaining $m-k$ components.

---

## 3. Subspace Reconstruction and Residual Error

In structural health monitoring, we often keep only the first $k$ principal components.

The centered observation vector can be decomposed as

$$
x_i-\hat{\mu}_n
=
\hat{P}_kz_{i,k}
+
\hat{P}_{m-k}z_{i,m-k}.
$$

Here,

$$
\hat{P}_kz_{i,k}
$$

is the explained principal subspace component, and

$$
\hat{P}_{m-k}z_{i,m-k}
$$

is the residual subspace component.

The reconstructed vector is

$$
\hat{x}_i
=
\hat{\mu}_n
+
\hat{P}_kz_{i,k}.
$$

The residual error vector is

$$
e_i
=
x_i-\hat{x}_i.
$$

Substituting $\hat{x}_i$,

$$
e_i
=
x_i-
(\hat{\mu}_n+\hat{P}_kz_{i,k}).
$$

Since

$$
x_i-\hat{\mu}_n
=
\hat{P}_kz_{i,k}
+
\hat{P}_{m-k}z_{i,m-k},
$$

we get

$$
e_i
=
\hat{P}_{m-k}z_{i,m-k}.
$$

---

## Pythagorean Energy Decomposition

Since the principal subspace and residual subspace are orthogonal,

$$
\hat{P}_k^T\hat{P}_{m-k}=0.
$$

Therefore, the total squared Euclidean norm decomposes as

$$
\|x_i-\hat{\mu}_n\|^2
=
\|\hat{P}_kz_{i,k}\|^2
+
\|\hat{P}_{m-k}z_{i,m-k}\|^2.
$$

Because $\hat{P}_k$ and $\hat{P}_{m-k}$ are orthonormal bases,

$$
\|\hat{P}_kz_{i,k}\|^2
=
\|z_{i,k}\|^2
$$

and

$$
\|\hat{P}_{m-k}z_{i,m-k}\|^2
=
\|z_{i,m-k}\|^2.
$$

Thus,

$$
\|x_i-\hat{\mu}_n\|^2
=
\|z_{i,k}\|^2
+
\|z_{i,m-k}\|^2.
$$

This proves that the total energy of the centered observation is split into explained principal energy and unexplained residual energy.

---

## Hotelling's $T^2$ Statistic

Hotelling's $T^2$ statistic measures the distance of an observation inside the retained principal subspace.

It is defined as

$$
T_i^2
=
z_{i,k}^T
\hat{\Lambda}_k^{-1}
z_{i,k}.
$$

Equivalently,

$$
T_i^2
=
\sum_{j=1}^{k}
\frac{z_{i,j}^2}{\hat{\lambda}_j}.
$$

A large value of $T_i^2$ means that the observation is unusual within the normal major variation directions.

In structural health monitoring, $T^2$ is sensitive to abnormal changes that still occur inside the normal principal subspace.

For example, an environmental load anomaly such as extreme operational wind loads may increase the energy inside the main structural modes. Therefore, it is usually detected strongly by

$$
T^2.
$$

---

## Residual $Q$ Statistic / Squared Prediction Error

The residual $Q$ statistic, also called Squared Prediction Error, measures the energy left outside the retained principal subspace.

It is defined as

$$
Q_i
=
e_i^Te_i.
$$

Since

$$
e_i = \hat{P}_{m-k}z_{i,m-k},
$$

we get

$$
Q_i
=
z_{i,m-k}^Tz_{i,m-k}.
$$

Equivalently,

$$
Q_i
=
\sum_{j=k+1}^{m}
z_{i,j}^2.
$$

A large $Q_i$ means that the observation contains information that the retained PCA model cannot explain.

In structural health monitoring, $Q$ is sensitive to new fault patterns or structural changes outside the learned healthy subspace.

For example, an internal structural fracture such as a fatigue crack can change the structural load pathway and create new residual behavior. Therefore, it is usually detected strongly by

$$
Q.
$$

---

## Final Diagnostic Comparison

Hotelling's $T^2$ statistic monitors abnormal variation inside the retained principal subspace.

The residual $Q$ statistic monitors abnormal variation outside the retained principal subspace.

Therefore:

$$
T^2
\rightarrow
\text{anomaly inside the normal dominant operating modes}
$$

and

$$
Q
\rightarrow
\text{new unmodeled anomaly or structural fault outside the normal subspace}.
$$

For an environmental load anomaly such as extreme wind loading, the more sensitive statistic is usually

$$
T^2.
$$

For an internal structural fracture or fatigue crack, the more sensitive statistic is usually

$$
Q.
$$

In [2]:
import numpy as np
import pandas as pd

# ------------------------------------------------------------
# Synthetic baseline data generator
# ------------------------------------------------------------
np.random.seed(42)

n_samples = 3000
n_features = 4

mean = np.zeros(n_features)

cov = np.array([
    [1.00, 0.75, 0.45, 0.20],
    [0.75, 1.00, 0.50, 0.25],
    [0.45, 0.50, 1.00, 0.60],
    [0.20, 0.25, 0.60, 1.00]
])

X = np.random.multivariate_normal(mean, cov, size=n_samples)

df_baseline = pd.DataFrame(
    X,
    columns=["Sensor_1", "Sensor_2", "Sensor_3", "Sensor_4"]
)

# ------------------------------------------------------------
# PCA Optimization Pipeline
# ------------------------------------------------------------
class PCASubspaceOptimizer:

    def __init__(self, df):
        self.df = df.select_dtypes(include=[np.number])

    def analyze_subspaces(self):
        X = self.df.values

        n, m = X.shape

        # 1. Center and standardize data
        X_mean = np.mean(X, axis=0)
        X_std = np.std(X, axis=0, ddof=1)

        X_std[X_std == 0] = 1e-15

        Z = (X - X_mean) / X_std

        # 2. Unbiased sample covariance matrix
        S = np.cov(Z, rowvar=False, ddof=1)

        # 3. Eigen decomposition
        eigenvalues, eigenvectors = np.linalg.eigh(S)

        # Sort eigenvalues and eigenvectors in descending order
        idx = np.argsort(eigenvalues)[::-1]
        eigenvalues = eigenvalues[idx]
        eigenvectors = eigenvectors[:, idx]

        # 4. PCA scores
        scores = Z @ eigenvectors

        results = []

        # 5. Loop through k = 1, 2, ..., m-1
        for k in range(1, m):

            scores_k = scores[:, :k]
            lambda_k = eigenvalues[:k]

            residual_scores = scores[:, k:]

            # Hotelling's T²
            T2_values = np.sum((scores_k ** 2) / lambda_k, axis=1)

            # Q Statistic / SPE
            Q_values = np.sum(residual_scores ** 2, axis=1)

            mean_T2 = np.mean(T2_values)
            mean_Q = np.mean(Q_values)

            explained_variance_ratio = np.sum(eigenvalues[:k]) / np.sum(eigenvalues)
            residual_variance_ratio = 1 - explained_variance_ratio

            results.append({
                "k": k,
                "Mean T2": mean_T2,
                "Mean Q": mean_Q,
                "Explained Variance Ratio": explained_variance_ratio,
                "Residual Variance Ratio": residual_variance_ratio
            })

        return pd.DataFrame(results), eigenvalues, eigenvectors


# ------------------------------------------------------------
# Execute PCA optimizer
# ------------------------------------------------------------
pca_engine = PCASubspaceOptimizer(df_baseline)

pca_results, eigenvalues, eigenvectors = pca_engine.analyze_subspaces()

print("Eigenvalues:")
print(eigenvalues)

print("\nPCA Subspace Optimization Results:")
display(pca_results)

Eigenvalues:
[2.40261148 1.0067027  0.34290837 0.24777746]

PCA Subspace Optimization Results:


,k,Mean T2,Mean Q,Explained Variance Ratio,Residual Variance Ratio
0,1,0.999667,1.596856,0.600653,0.399347
1,2,1.999333,0.590489,0.852329,0.147671
2,3,2.999000,0.247695,0.938056,0.061944


# Latent Subspace Decomposition via Factor Analysis (FA)

# Part A: Theoretical Fundamentals (Generative Model & Commonalities)

## 1. Fundamental Equation of Factor Analysis

Let

$$
Z_i \in \mathbb{R}^m
$$

be a normalized zero-mean baseline feature vector.

Since the data is normalized,

$$
E[Z_i]=0.
$$

The population covariance matrix of $Z_i$ is the correlation matrix

$$
R = E[Z_iZ_i^T].
$$

Factor Analysis assumes that the observed vector $Z_i$ is generated by a smaller number of hidden latent factors.

The Factor Analysis model is

$$
Z_i = \Lambda F_i + \epsilon_i.
$$

Here,

$$
F_i \in \mathbb{R}^k
$$

is the hidden latent factor vector, where

$$
k < m.
$$

The matrix

$$
\Lambda \in \mathbb{R}^{m \times k}
$$

is the factor loading matrix.

The vector

$$
\epsilon_i \in \mathbb{R}^m
$$

is the sensor-specific noise or unique error vector.

The assumptions are:

$$
E[F_i]=0,
$$

$$
E[F_iF_i^T]=I_{k \times k},
$$

$$
E[\epsilon_i]=0,
$$

$$
E[\epsilon_i\epsilon_i^T]=\Psi,
$$

where

$$
\Psi = diag(\phi_1^2,\phi_2^2,\ldots,\phi_m^2).
$$

Also, the latent factors and errors are independent:

$$
E[\epsilon_iF_i^T]=0.
$$

Now,

$$
R = E[Z_iZ_i^T].
$$

Substituting the factor model,

$$
R
=
E[(\Lambda F_i+\epsilon_i)(\Lambda F_i+\epsilon_i)^T].
$$

Expanding,

$$
R
=
E[\Lambda F_iF_i^T\Lambda^T]
+
E[\Lambda F_i\epsilon_i^T]
+
E[\epsilon_iF_i^T\Lambda^T]
+
E[\epsilon_i\epsilon_i^T].
$$

Since $\Lambda$ is constant,

$$
R
=
\Lambda E[F_iF_i^T]\Lambda^T
+
\Lambda E[F_i\epsilon_i^T]
+
E[\epsilon_iF_i^T]\Lambda^T
+
E[\epsilon_i\epsilon_i^T].
$$

Using the assumptions,

$$
E[F_iF_i^T]=I,
$$

$$
E[F_i\epsilon_i^T]=0,
$$

$$
E[\epsilon_iF_i^T]=0,
$$

and

$$
E[\epsilon_i\epsilon_i^T]=\Psi,
$$

we get

$$
R
=
\Lambda I\Lambda^T
+
0
+
0
+
\Psi.
$$

Therefore,

$$
R = \Lambda\Lambda^T+\Psi.
$$

This is the Fundamental Equation of Factor Analysis.

It means that the observed correlation structure is divided into two parts:

$$
R
=
\text{shared structural variance}
+
\text{unique sensor noise}.
$$

The shared structural part is

$$
\Lambda\Lambda^T,
$$

and the unique noise part is

$$
\Psi.
$$

---

## Communality and Uniqueness

Now consider the $j$-th diagonal element of the matrix equation

$$
R = \Lambda\Lambda^T+\Psi.
$$

Since $Z_i$ is standardized, each variable has unit variance:

$$
R_{jj}=1.
$$

The $j$-th diagonal element becomes

$$
1
=
\sum_{r=1}^{k}\lambda_{j,r}^2
+
\phi_j^2.
$$

The first term is called the communality of the $j$-th sensor:

$$
h_j^2
=
\sum_{r=1}^{k}\lambda_{j,r}^2.
$$

The second term is called uniqueness:

$$
\phi_j^2
=
1-h_j^2.
$$

Therefore,

$$
1=h_j^2+\phi_j^2.
$$

---

## Physical Interpretation

The communality

$$
h_j^2
$$

represents the portion of the sensor variance explained by the common latent factors.

If

$$
h_j^2 \approx 1,
$$

then the sensor mainly captures shared structural behavior.

This means the sensor is strongly connected to the actual physical system.

The uniqueness

$$
\phi_j^2
$$

represents the portion of variance that is specific only to that sensor.

If

$$
\phi_j^2 \approx 1,
$$

then the sensor is mostly measuring isolated noise or sensor-specific disturbance.

In an asset diagnostics system:

$$
h_j^2 \text{ high}
\Rightarrow
\text{sensor is structurally meaningful}
$$

and

$$
\phi_j^2 \text{ high}
\Rightarrow
\text{sensor may be noisy, faulty, or poorly calibrated}.
$$

---

## 2. PCA Loading Vector vs Varimax-Rotated Factor Loading Vector

A raw PCA loading vector is obtained by finding directions that maximize total variance.

PCA focuses on the total variance in the data.

Therefore, PCA may mix structural signal and sensor noise together if both contribute large variance.

In PCA, the principal components are ordered as

$$
\lambda_1 \geq \lambda_2 \geq \cdots \geq \lambda_m.
$$

The first principal component captures the largest possible variance, the second captures the next largest variance, and so on.

However, Factor Analysis tries to explain the correlation structure using hidden common factors.

In Factor Analysis, the loading matrix

$$
\Lambda
$$

shows how strongly each sensor is connected to each latent physical factor.

A Varimax rotation is often applied to make the factor loading matrix easier to interpret.

The goal of Varimax rotation is to create a simple structure.

This means each sensor should load strongly on only one factor and weakly on the others.

For example, before rotation a loading matrix may look like

$$
\Lambda =
\begin{bmatrix}
0.60 & 0.55 \\
0.58 & 0.52 \\
0.50 & 0.57
\end{bmatrix}.
$$

This is difficult to interpret because every sensor loads on every factor.

After Varimax rotation, it may become

$$
\Lambda_{rotated} =
\begin{bmatrix}
0.85 & 0.05 \\
0.82 & 0.10 \\
0.08 & 0.90
\end{bmatrix}.
$$

Now interpretation is easier.

Sensor 1 and Sensor 2 mainly belong to Factor 1.

Sensor 3 mainly belongs to Factor 2.

---

## Why Rotation Does Not Change Communality

Let $T$ be an orthogonal rotation matrix.

Then the rotated loading matrix is

$$
\Lambda_{rotated}=\Lambda T.
$$

Because $T$ is orthogonal,

$$
TT^T=I.
$$

Now the common covariance after rotation is

$$
\Lambda_{rotated}\Lambda_{rotated}^T.
$$

Substituting,

$$
\Lambda_{rotated}\Lambda_{rotated}^T
=
(\Lambda T)(\Lambda T)^T.
$$

Using the transpose rule,

$$
(\Lambda T)^T=T^T\Lambda^T.
$$

Therefore,

$$
\Lambda_{rotated}\Lambda_{rotated}^T
=
\Lambda TT^T\Lambda^T.
$$

Since

$$
TT^T=I,
$$

we get

$$
\Lambda_{rotated}\Lambda_{rotated}^T
=
\Lambda I\Lambda^T.
$$

Thus,

$$
\Lambda_{rotated}\Lambda_{rotated}^T
=
\Lambda\Lambda^T.
$$

Therefore, rotation changes the appearance of the loading matrix but does not change the total shared covariance structure.

Because communality is calculated as

$$
h_j^2=\sum_{r=1}^{k}\lambda_{j,r}^2,
$$

orthogonal rotation preserves the row-wise sum of squared loadings.

Therefore,

$$
h_{j,rotated}^2=h_j^2.
$$

So, Varimax rotation improves interpretability without changing the communality of each sensor or the global covariance approximation

$$
R \approx \Lambda\Lambda^T+\Psi.
$$

---

## 3. Thomson's Regression Method for Factor Score Estimation

The latent factors

$$
F_i
$$

are hidden and cannot be directly observed.

Unlike PCA, we cannot obtain factor scores using a simple geometric projection.

Instead, we estimate the factor score using conditional expectation:

$$
f_i=E[F_i|Z_i=z_i].
$$

This is called Thomson's Regression Method.

It is also a Minimum Mean Squared Error estimator.

---

## Constructing the Joint Vector

Define the joint vector

$$
Y_i=
\begin{bmatrix}
Z_i \\
F_i
\end{bmatrix}.
$$

Since

$$
E[Z_i]=0
$$

and

$$
E[F_i]=0,
$$

we have

$$
E[Y_i]=0.
$$

Now construct the covariance matrix of $Y_i$:

$$
\Sigma_{YY}
=
E[Y_iY_i^T].
$$

Therefore,

$$
\Sigma_{YY}
=
\begin{bmatrix}
E[Z_iZ_i^T] & E[Z_iF_i^T] \\
E[F_iZ_i^T] & E[F_iF_i^T]
\end{bmatrix}.
$$

From Factor Analysis,

$$
E[Z_iZ_i^T]=R.
$$

Also,

$$
E[F_iF_i^T]=I_{k \times k}.
$$

Now calculate

$$
E[Z_iF_i^T].
$$

Using

$$
Z_i=\Lambda F_i+\epsilon_i,
$$

we get

$$
E[Z_iF_i^T]
=
E[(\Lambda F_i+\epsilon_i)F_i^T].
$$

Expanding,

$$
E[Z_iF_i^T]
=
\Lambda E[F_iF_i^T]
+
E[\epsilon_iF_i^T].
$$

Using

$$
E[F_iF_i^T]=I
$$

and

$$
E[\epsilon_iF_i^T]=0,
$$

we get

$$
E[Z_iF_i^T]=\Lambda.
$$

Therefore,

$$
E[F_iZ_i^T]=\Lambda^T.
$$

So the joint covariance matrix becomes

$$
\Sigma_{YY}
=
\begin{bmatrix}
R & \Lambda \\
\Lambda^T & I_{k \times k}
\end{bmatrix}.
$$

---

## Conditional Expectation Formula

For a joint normal system,

$$
\begin{bmatrix}
X_1 \\
X_2
\end{bmatrix}
$$

with covariance matrix

$$
\begin{bmatrix}
\Sigma_{11} & \Sigma_{12} \\
\Sigma_{21} & \Sigma_{22}
\end{bmatrix},
$$

the conditional expectation is

$$
E[X_2|X_1=x_1]
=
\mu_2
+
\Sigma_{21}\Sigma_{11}^{-1}
(x_1-\mu_1).
$$

For Factor Analysis, we set

$$
X_1=Z_i
$$

and

$$
X_2=F_i.
$$

Then,

$$
\mu_1=0,
$$

$$
\mu_2=0,
$$

$$
\Sigma_{21}=\Lambda^T,
$$

and

$$
\Sigma_{11}=R.
$$

Substituting into the conditional expectation formula,

$$
E[F_i|Z_i=z_i]
=
0+\Lambda^TR^{-1}(z_i-0).
$$

Therefore,

$$
f_i=\Lambda^TR^{-1}z_i.
$$

This is Thomson's factor score estimator.



#Part B: Parametric Factor Partition Engine & UI Architecture

In [1]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.decomposition import FactorAnalysis

# -----------------------------
# Synthetic data from assignment
# -----------------------------
np.random.seed(42)
n_samples = 2500

f1 = np.random.normal(0, 1, n_samples)
f2 = np.random.normal(0, 1, n_samples)

s1 = 0.85 * f1 + 0.10 * f2 + np.random.normal(0, 0.3, n_samples)
s2 = 0.80 * f1 + 0.15 * f2 + np.random.normal(0, 0.35, n_samples)
s3 = 0.12 * f1 + 0.90 * f2 + np.random.normal(0, 0.25, n_samples)
s4 = 0.02 * f1 + 0.05 * f2 + np.random.normal(0, 1.40, n_samples)

df_asset = pd.DataFrame(
    np.vstack([s1, s2, s3, s4]).T,
    columns=["Sensor_1", "Sensor_2", "Sensor_3", "Sensor_4"]
)

# -----------------------------
# Varimax rotation
# -----------------------------
def varimax(Phi, gamma=1.0, q=50, tol=1e-6):
    p, k = Phi.shape
    R = np.eye(k)
    d_old = 0

    for _ in range(q):
        Lambda = Phi @ R
        u, s, vh = np.linalg.svd(
            Phi.T @ (Lambda**3 - (gamma / p) * Lambda @ np.diag(np.diag(Lambda.T @ Lambda)))
        )
        R = u @ vh
        d_new = np.sum(s)

        if d_old != 0 and d_new / d_old < 1 + tol:
            break

        d_old = d_new

    return Phi @ R


# -----------------------------
# Factor Analysis Engine
# -----------------------------
class FactorPartitionEngine:
    def __init__(self, n_factors=2):
        self.n_factors = n_factors

    def fit_transform(self, df):
        numeric_df = df.select_dtypes(include=[np.number])

        Z_raw = numeric_df.values
        n, m = Z_raw.shape

        if n <= m:
            raise ValueError("Number of observations n must be greater than number of features m.")

        if self.n_factors >= m:
            raise ValueError("Number of latent factors k must be strictly less than number of features m.")

        feature_names = numeric_df.columns.tolist()

        mean = np.mean(Z_raw, axis=0)
        std = np.std(Z_raw, axis=0, ddof=1)

        std[std == 0] = 1e-15

        Z = (Z_raw - mean) / std

        R = np.cov(Z, rowvar=False, ddof=1)

        fa = FactorAnalysis(n_components=self.n_factors, random_state=42)
        fa.fit(Z)

        loadings = fa.components_.T

        rotated_loadings = varimax(loadings)

        communality = np.sum(rotated_loadings**2, axis=1)

        uniqueness = 1 - communality
        uniqueness = np.clip(uniqueness, 0, 1)

        factor_scores = Z @ np.linalg.inv(R) @ rotated_loadings

        factor_variance = np.var(factor_scores, axis=0, ddof=1)

        print("Average System Communality %:", np.mean(communality) * 100)
        print("Average System Uniqueness %:", np.mean(uniqueness) * 100)

        return {
            "Z": Z,
            "R": R,
            "loadings": loadings,
            "rotated_loadings": rotated_loadings,
            "communality": communality,
            "uniqueness": uniqueness,
            "factor_scores": factor_scores,
            "factor_variance": factor_variance,
            "feature_names": feature_names
        }


engine = FactorPartitionEngine(n_factors=2)
results = engine.fit_transform(df_asset)

rotated_loadings = results["rotated_loadings"]
communality = results["communality"]
uniqueness = results["uniqueness"]
factor_variance = results["factor_variance"]
feature_names = results["feature_names"]

factor_names = [f"Factor_{i+1}" for i in range(rotated_loadings.shape[1])]

# -----------------------------
# 2 x 2 Dashboard
# -----------------------------
fig = make_subplots(
    rows=2,
    cols=2,
    horizontal_spacing=0.24,
    vertical_spacing=0.28,
    subplot_titles=(
        "Structural Loadings Matrix",
        "Variance Partitioning Profile",
        "Sensor Uniqueness Noise Floor",
        "Latent Factor Variance"
    )
)

# 1. Structural loadings heatmap
fig.add_trace(
    go.Heatmap(
        z=np.abs(rotated_loadings),
        x=factor_names,
        y=feature_names,
        colorscale="YlOrRd",
        colorbar=dict(
            title="|Loading|",
            x=-0.15,
            len=0.38,
            y=0.78,
            yanchor="middle",
            xanchor="right"
        ),
        text=np.round(np.abs(rotated_loadings), 3),
        texttemplate="%{text}"
    ),
    row=1,
    col=1
)

# 2. Horizontal stacked bar chart
fig.add_trace(
    go.Bar(
        y=feature_names,
        x=communality * 100,
        orientation="h",
        name="Shared Communality h²",
        marker=dict(color="#1f77b4")
    ),
    row=1,
    col=2
)

fig.add_trace(
    go.Bar(
        y=feature_names,
        x=uniqueness * 100,
        orientation="h",
        name="Uniqueness φ²",
        marker=dict(color="#ff7f0e")
    ),
    row=1,
    col=2
)

# 3. Sensor uniqueness line
fig.add_trace(
    go.Scatter(
        x=feature_names,
        y=uniqueness,
        mode="lines+markers",
        name="Uniqueness φ²",
        line=dict(color="#d62728", width=2, dash="dashdot")
    ),
    row=2,
    col=1
)

# 4. Latent factor variance bar chart
fig.add_trace(
    go.Bar(
        x=factor_names,
        y=factor_variance,
        name="Factor Variance",
        marker=dict(color="#2ca02c", line=dict(color="black", width=0.5))
    ),
    row=2,
    col=2
)

fig.update_layout(
    title="Factor Analysis Diagnostic Dashboard",
    template="plotly_white",
    width=1250,
    height=750,
    barmode="stack",
    legend=dict(orientation="h", x=0.5, y=1.02, xanchor="center"),
    margin=dict(t=150, b=60, l=140, r=80)
)

fig.update_xaxes(range=[0, 100], row=1, col=2)
fig.update_xaxes(tickangle=25, row=2, col=1)

fig.show()

# -----------------------------
# Numerical summary table
# -----------------------------
summary_df = pd.DataFrame({
    "Sensor": feature_names,
    "Communality_h2": communality,
    "Uniqueness_phi2": uniqueness
})

display(summary_df)

factor_variance_df = pd.DataFrame({
    "Factor": factor_names,
    "Empirical_Variance": factor_variance
})

display(factor_variance_df)

Average System Communality %: 50.6586398197277
Average System Uniqueness %: 49.3413601802723


,Sensor,Communality_h2,Uniqueness_phi2
0,Sensor_1,0.864586,0.135414
1,Sensor_2,0.875204,0.124796
2,Sensor_3,0.277897,0.722103
3,Sensor_4,0.008658,0.991342


,Factor,Empirical_Variance
0,Factor_1,0.925003
1,Factor_2,0.284073


##Explanation

The Factor Analysis engine first standardizes the sensor data using

$$
Z_{i,j}=\frac{x_{i,j}-\hat{\mu}_j}{s_j}
$$

where $s_j$ is calculated using

$$
ddof=1.
$$

The standardized data matrix is

$$
Z \in \mathbb{R}^{n \times m}.
$$

The empirical correlation matrix is calculated as

$$
R=\frac{1}{n-1}Z^TZ.
$$

A Factor Analysis model with

$$
k=2
$$

latent factors is fitted to decompose the sensor structure into

$$
R \approx \Lambda\Lambda^T+\Psi.
$$

The loading matrix is then rotated using Varimax rotation to improve interpretability.

The communality of each sensor is calculated as

$$
h_j^2=\sum_{r=1}^{k}\lambda_{j,r}^2.
$$

The uniqueness is calculated as

$$
\phi_j^2=1-h_j^2.
$$

The latent factor scores are estimated using Thomson's regression method:

$$
F=ZR^{-1}\Lambda_{rotated}.
$$

The dashboard shows:

1. The absolute rotated loading matrix $|\Lambda_{rotated}|$.
2. The variance partitioning between $h_j^2$ and $\phi_j^2$.
3. The sensor uniqueness noise floor.
4. The empirical variance of the extracted latent factor scores.

A sensor with high communality is strongly explained by shared structural factors.

A sensor with high uniqueness is mostly dominated by isolated sensor noise or local measurement fault.

#Subspace Diagnostics & Feature De-correlation
##Part A : Develop a Feature Analysis Dashboard

In [2]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.decomposition import FactorAnalysis

# -----------------------------
# Synthetic data
# -----------------------------
np.random.seed(42)
n_samples = 2500

f1 = np.random.normal(0, 1, n_samples)
f2 = np.random.normal(0, 1, n_samples)

s1 = 0.85*f1 + 0.10*f2 + np.random.normal(0, 0.3, n_samples)
s2 = 0.80*f1 + 0.15*f2 + np.random.normal(0, 0.35, n_samples)
s3 = 0.12*f1 + 0.90*f2 + np.random.normal(0, 0.25, n_samples)
s4 = 0.02*f1 + 0.05*f2 + np.random.normal(0, 1.40, n_samples)

df_asset = pd.DataFrame(
    np.vstack([s1, s2, s3, s4]).T,
    columns=["Sensor_1", "Sensor_2", "Sensor_3", "Sensor_4"]
)

# -----------------------------
# Standardization
# -----------------------------
X = df_asset.values
feature_names = df_asset.columns.tolist()

X_mean = X.mean(axis=0)
X_std = X.std(axis=0, ddof=1)
X_std[X_std == 0] = 1e-15

Z = (X - X_mean) / X_std
n, m = Z.shape

# -----------------------------
# PCA engine
# -----------------------------
S = np.cov(Z, rowvar=False, ddof=1)

eigvals, eigvecs = np.linalg.eigh(S)
idx = np.argsort(eigvals)[::-1]

eigvals = eigvals[idx]
eigvecs = eigvecs[:, idx]

scores = Z @ eigvecs

explained_var = eigvals / eigvals.sum()
cumulative_var = np.cumsum(explained_var)
residual_var = 1 - cumulative_var

k_values = np.arange(1, m)

mean_T2 = []
mean_Q = []

for k in k_values:
    scores_k = scores[:, :k]
    lambdas_k = eigvals[:k]
    residual_scores = scores[:, k:]

    T2 = np.sum((scores_k ** 2) / lambdas_k, axis=1)
    Q = np.sum(residual_scores ** 2, axis=1)

    mean_T2.append(np.mean(T2))
    mean_Q.append(np.mean(Q))

# -----------------------------
# PCA dashboard
# -----------------------------
fig_pca = make_subplots(
    rows=2,
    cols=3,
    subplot_titles=(
        "PCA Loadings Matrix |P|",
        "Absolute Eigenvalues",
        "Explained Variance",
        "Residual Space %",
        "Mean Hotelling T² vs k",
        "Mean Q/SPE vs k"
    ),
    horizontal_spacing=0.08,
    vertical_spacing=0.22
)

fig_pca.add_trace(
    go.Heatmap(
        z=np.abs(eigvecs),
        x=[f"PC{i+1}" for i in range(m)],
        y=feature_names,
        colorscale="YlOrRd",
        text=np.round(np.abs(eigvecs), 3),
        texttemplate="%{text}"
    ),
    row=1,
    col=1
)

fig_pca.add_trace(
    go.Bar(
        x=[f"PC{i+1}" for i in range(m)],
        y=eigvals,
        name="Eigenvalues"
    ),
    row=1,
    col=2
)

fig_pca.add_trace(
    go.Bar(
        x=[f"PC{i+1}" for i in range(m)],
        y=explained_var * 100,
        name="Marginal Explained %"
    ),
    row=1,
    col=3
)

fig_pca.add_trace(
    go.Scatter(
        x=[f"PC{i+1}" for i in range(m)],
        y=cumulative_var * 100,
        mode="lines+markers",
        name="Cumulative Explained %",
        line=dict(dash="dash")
    ),
    row=1,
    col=3
)

fig_pca.add_trace(
    go.Bar(
        x=[f"PC{i+1}" for i in range(m)],
        y=residual_var * 100,
        name="Residual %"
    ),
    row=2,
    col=1
)

fig_pca.add_trace(
    go.Scatter(
        x=k_values,
        y=mean_T2,
        mode="lines+markers",
        name="Mean T²"
    ),
    row=2,
    col=2
)

fig_pca.add_trace(
    go.Scatter(
        x=k_values,
        y=mean_Q,
        mode="lines+markers",
        name="Mean Q"
    ),
    row=2,
    col=3
)

fig_pca.update_layout(
    title="PCA Optimization Suite",
    template="plotly_white",
    width=1350,
    height=750,
    showlegend=True
)

fig_pca.show()

# -----------------------------
# Varimax rotation
# -----------------------------
def varimax(Phi, gamma=1.0, q=50, tol=1e-6):
    p, k = Phi.shape
    R = np.eye(k)
    d_old = 0

    for _ in range(q):
        Lambda = Phi @ R
        u, s, vh = np.linalg.svd(
            Phi.T @ (
                Lambda**3
                - (gamma / p) * Lambda @ np.diag(np.diag(Lambda.T @ Lambda))
            )
        )
        R = u @ vh
        d_new = np.sum(s)

        if d_old != 0 and d_new / d_old < 1 + tol:
            break

        d_old = d_new

    return Phi @ R

# -----------------------------
# FA engine
# -----------------------------
k_factors = 2

fa = FactorAnalysis(n_components=k_factors, random_state=42)
fa.fit(Z)

loadings = fa.components_.T
rotated_loadings = varimax(loadings)

communality = np.sum(rotated_loadings**2, axis=1)
uniqueness = 1 - communality
uniqueness = np.clip(uniqueness, 0, 1)

R_corr = np.cov(Z, rowvar=False, ddof=1)
factor_scores = Z @ np.linalg.inv(R_corr) @ rotated_loadings
factor_variance = np.var(factor_scores, axis=0, ddof=1)

factor_names = [f"Factor_{i+1}" for i in range(k_factors)]

# -----------------------------
# FA dashboard
# -----------------------------
fig_fa = make_subplots(
    rows=2,
    cols=2,
    subplot_titles=(
        "FA Structural Loadings Matrix",
        "Variance Allocation: h² vs φ²",
        "Sensor Uniqueness Noise Floor",
        "Latent Factor Score Variance"
    ),
    horizontal_spacing=0.24,
    vertical_spacing=0.28
)

fig_fa.add_trace(
    go.Heatmap(
        z=np.abs(rotated_loadings),
        x=factor_names,
        y=feature_names,
        colorscale="YlOrRd",
        text=np.round(np.abs(rotated_loadings), 3),
        texttemplate="%{text}",
        colorbar=dict(
            title="|Loading|",
            x=-0.15,
            len=0.38,
            y=0.78,
            yanchor="middle",
            xanchor="right"
        )
    ),
    row=1,
    col=1
)

fig_fa.add_trace(
    go.Bar(
        y=feature_names,
        x=communality * 100,
        orientation="h",
        name="Shared Communality h²",
        marker=dict(color="#1f77b4")
    ),
    row=1,
    col=2
)

fig_fa.add_trace(
    go.Bar(
        y=feature_names,
        x=uniqueness * 100,
        orientation="h",
        name="Uniqueness φ²",
        marker=dict(color="#ff7f0e")
    ),
    row=1,
    col=2
)

fig_fa.add_trace(
    go.Scatter(
        x=feature_names,
        y=uniqueness,
        mode="lines+markers",
        name="Uniqueness φ²",
        line=dict(color="#d62728", width=2, dash="dashdot")
    ),
    row=2,
    col=1
)

fig_fa.add_trace(
    go.Bar(
        x=factor_names,
        y=factor_variance,
        name="Latent Factor Variance",
        marker=dict(color="#2ca02c", line=dict(color="black", width=0.5))
    ),
    row=2,
    col=2
)

fig_fa.update_layout(
    title="FA Latent Subspace Suite",
    template="plotly_white",
    width=1250,
    height=750,
    barmode="stack",
    legend=dict(orientation="h", x=0.5, y=1.02, xanchor="center"),
    margin=dict(t=150, b=60, l=140, r=80)
)

fig_fa.update_xaxes(range=[0, 100], row=1, col=2)
fig_fa.update_xaxes(tickangle=25, row=2, col=1)

fig_fa.show()

# -----------------------------
# Summary tables
# -----------------------------
pca_summary = pd.DataFrame({
    "PC": [f"PC{i+1}" for i in range(m)],
    "Eigenvalue": eigvals,
    "Marginal Explained Variance %": explained_var * 100,
    "Cumulative Explained Variance %": cumulative_var * 100,
    "Residual Variance %": residual_var * 100
})

fa_summary = pd.DataFrame({
    "Sensor": feature_names,
    "Communality_h2": communality,
    "Uniqueness_phi2": uniqueness
})

display(pca_summary)
display(fa_summary)

,PC,Eigenvalue,Marginal Explained Variance %,Cumulative Explained Variance %,Residual Variance %
0,PC1,1.974420,49.360512,49.360512,5.063949e+01
1,PC2,1.006968,25.174209,74.534721,2.546528e+01
2,PC3,0.881088,22.027198,96.561920,3.438080e+00
3,PC4,0.137523,3.438080,100.000000,1.110223e-14


,Sensor,Communality_h2,Uniqueness_phi2
0,Sensor_1,0.864586,0.135414
1,Sensor_2,0.875204,0.124796
2,Sensor_3,0.277897,0.722103
3,Sensor_4,0.008658,0.991342


# Part B: Analysis

## Question 1: The Total Variance Illusion

### 1. Meaning of Sensor_4 uniqueness close to 100%

From the FA result,

$$
\phi_4^2 = 0.991342
$$

or approximately

$$
99.13\%.
$$

This means Sensor_4 is almost completely dominated by unique sensor-specific noise. Its communality is only

$$
h_4^2 = 0.008658
$$

or approximately

$$
0.87\%.
$$

Therefore, Sensor_4 does not significantly represent the shared physical structure of the asset. It is likely affected by high localized noise, calibration error, or instrumentation fault.

---

### 2. How Sensor_4 tricks PCA

PCA maximizes total variance without knowing whether the variance is useful structural variance or useless sensor noise.

From the PCA result,

$$
PC_1 = 49.36\%
$$

and

$$
PC_2 = 25.17\%.
$$

Together,

$$
PC_1 + PC_2 = 74.53\%.
$$

However, Sensor_4 has a very high loading on PC2:

$$
|P_{Sensor4,PC2}| = 0.95.
$$

This means PC2 is strongly influenced by Sensor_4. Since Sensor_4 is mostly noise according to FA, PCA incorrectly treats this noisy variance as an important principal direction.

---

### 3. Risk of relying only on PCA

If an engineer uses PCA alone, Sensor_4 noise may be treated as real structural information.

This creates major risks:

- False alarms may increase.
- True structural changes may be hidden.
- The PCA eigenvalues may become inflated by noise.
- Monitoring thresholds may be built around faulty sensor behavior.

Therefore, PCA alone can be misleading when one sensor has very high localized noise.

---

## Question 2: Decoupling Structural Loading via Rotation

### 1. How Varimax creates simple structure

PCA follows the eigenvalue hierarchy

$$
\lambda_1 > \lambda_2 > \lambda_3 > \lambda_4.
$$

Each principal component tries to capture maximum total variance. Therefore, PCA components can mix useful structural variance and sensor noise.

In the PCA loading matrix, Sensor_4 strongly loads on PC2:

$$
|P_{Sensor4,PC2}| = 0.95.
$$

This shows that PCA gives high importance to Sensor_4 because of its large variance.

In Factor Analysis, Varimax rotation produces a simpler and more interpretable structure. From the FA loading matrix:

$$
Sensor_1 \rightarrow Factor_1 = 0.930
$$

$$
Sensor_2 \rightarrow Factor_1 = 0.924
$$

$$
Sensor_3 \rightarrow Factor_2 = 0.482
$$

Sensor_1 and Sensor_2 clearly belong to Factor_1. Sensor_3 is more related to Factor_2. Sensor_4 has very weak loadings on both factors.

This is called simple structure.

---

### 2. Why FA heatmap is easier for troubleshooting

The FA heatmap is easier for a plant operator because it separates the sensors according to hidden physical factors.

From the results:

- Sensor_1 and Sensor_2 are strongly connected to Factor_1.
- Sensor_3 is mainly connected to Factor_2.
- Sensor_4 is mostly noise.

Therefore, if Factor_1 changes, the operator can focus on the physical system measured by Sensor_1 and Sensor_2.

If Sensor_4 changes, the operator should suspect sensor noise or sensor malfunction rather than structural damage.

PCA does not provide this clear physical separation.

---

## Question 3: Determining Subspace Truncation Using $T^2$ and $Q$

### 1. Behavior of the Mean $Q$ curve

The residual statistic is

$$
Q_i = \sum_{j=k+1}^{m}z_{i,j}^2.
$$

From the PCA result, the Mean $Q$ curve decreases as $k$ increases.

When

$$
k=1 \rightarrow k=2,
$$

Mean $Q$ drops sharply from about

$$
2.0
$$

to about

$$
1.0.
$$

This means the second principal component captures a large amount of previously unexplained energy.

When

$$
k=2 \rightarrow k=3,
$$

Mean $Q$ drops again to about

$$
0.14.
$$

After this point, only a very small residual remains.

---

### 2. Why the elbow identifies hidden dimensionality

The cumulative explained variance values are:

$$
PC_1 = 49.36\%
$$

$$
PC_1+PC_2 = 74.53\%
$$

$$
PC_1+PC_2+PC_3 = 96.56\%.
$$

Since the residual variance becomes very small after

$$
k=3,
$$

the practical elbow appears at

$$
k=3.
$$

This means the PCA model needs three components to capture almost all system variance.

However, FA shows that one of the components is likely influenced by Sensor_4 noise. Therefore, physically meaningful hidden structure is mainly represented by two latent factors, while the third PCA component mostly captures remaining variance/noise.

---

### 3. What happens if $k=3$ is chosen incorrectly?

If

$$
k=3
$$

is selected as the clean principal subspace, the model includes almost all variance:

$$
96.56\%.
$$

However, because Sensor_4 is noisy, part of this retained variance is not structural.

Thus, choosing

$$
k=3
$$

forces noisy sensor behavior into the clean principal subspace.

This may reduce the sensitivity of anomaly detection because the model begins to accept noise as normal behavior.

---

## Question 4: Operational Trade-offs in System Health Monitoring

### PCA Strategy

The PCA strategy monitors:

$$
T^2
$$

inside the retained principal subspace and

$$
Q
$$

in the residual subspace.

PCA is useful because it compresses data and removes correlation.

However, PCA cannot separate structural variance from sensor noise.

In this result, Sensor_4 has a high PCA loading on PC2:

$$
|P_{Sensor4,PC2}| = 0.95.
$$

So PCA gives high importance to Sensor_4 even though FA shows it is mostly noise.

---

### FA Strategy

The FA strategy models

$$
R \approx \Lambda\Lambda^T+\Psi.
$$

It separates shared structural variance from unique sensor noise.

From the FA result:

$$
\phi_1^2 = 0.135414
$$

$$
\phi_2^2 = 0.124796
$$

$$
\phi_3^2 = 0.722103
$$

$$
\phi_4^2 = 0.991342
$$

Sensor_4 has the highest uniqueness, almost equal to 1.

Therefore, FA clearly identifies Sensor_4 as a faulty or noisy sensor.

